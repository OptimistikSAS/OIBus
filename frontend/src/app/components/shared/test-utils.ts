import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
