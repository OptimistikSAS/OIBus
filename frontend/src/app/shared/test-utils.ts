import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
