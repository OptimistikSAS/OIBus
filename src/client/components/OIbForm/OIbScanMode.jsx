import React from 'react'
import PropTypes from 'prop-types'
import { FormGroup, FormText, Label, Input, FormFeedback } from 'reactstrap'
import { ConfigContext } from '../../context/configContext.jsx'

const OIbScanMode = ({ label, help, value, name, onChange }) => {
  const { newConfig } = React.useContext(ConfigContext)
  const scanModes = newConfig?.engine?.scanModes ?? [] // scan modes defined in engine
  let options = scanModes.map((e) => e.scanMode)
  if (options === null || options.length === 0) {
    options = [''] // empty string if no scan mode on engine
  }
  const defaultValue = options[0]
  let validCheck = null

  React.useEffect(() => {
    if (value === '') onChange(name, defaultValue)
  }, [value])

  React.useEffect(() => {
    // save error if validCheck has error message
    onChange(name, value, validCheck)
  }, [validCheck])

  const handleChange = (event) => {
    const { target } = event
    const { value: newVal } = target
    onChange(name, newVal, null)
  }

  // check if defined scanmode is unknown to the engine
  if (!options.includes(value)) {
    options.unshift(value)
    validCheck = 'Invalid scan mode'
  }

  return (
    <FormGroup>
      {label && <Label for={name}>{label}</Label>}
      <Input className="oi-form-input" type="select" id={name} name={name} invalid={validCheck !== null} onChange={handleChange} value={value}>
        {options.map((o) => (
          <option key={o} value={o}>
            {o}
          </option>
        ))}
      </Input>
      <FormFeedback>{validCheck}</FormFeedback>
      {help && <FormText>{help}</FormText>}
    </FormGroup>
  )
}
OIbScanMode.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  help: PropTypes.element,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
}
OIbScanMode.defaultProps = { label: null, help: null, value: '' }

export default OIbScanMode
