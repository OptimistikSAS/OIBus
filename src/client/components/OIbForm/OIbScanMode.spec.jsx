import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import OIbScanMode from './OIbScanMode.jsx'

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('OIbScanMode', () => {
  test('check ScanMode with value="value"', () => {
    act(() => {
      ReactDOM.render(<OIbScanMode
        name="name"
        label="label"
        value="value"
        onChange={() => (1)}
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
  test('check ScanMode with no label', () => {
    act(() => {
      ReactDOM.render(<OIbScanMode
        name="name"
        value="value"
        onChange={() => (1)}
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
  test('check ScanMode with value="value" and help', () => {
    act(() => {
      ReactDOM.render(<OIbScanMode
        name="name"
        help={<div>help text</div>}
        label="label"
        value="value"
        onChange={() => (1)}
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
})
