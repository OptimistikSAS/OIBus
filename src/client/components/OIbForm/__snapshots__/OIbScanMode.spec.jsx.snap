// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OIbScanMode check ScanMode with no label 1`] = `
<div>
  <div
    class="form-group"
  >
    <select
      class="oi-form-input is-invalid form-control"
      id="name"
      name="name"
    >
      <option
        value="value"
      >
        value
      </option>
      <option
        value=""
      />
    </select>
    <div
      class="invalid-feedback"
    >
      Invalid scan mode
    </div>
  </div>
</div>
`;

exports[`OIbScanMode check ScanMode with value="value" 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class=""
      for="name"
    >
      label
    </label>
    <select
      class="oi-form-input is-invalid form-control"
      id="name"
      name="name"
    >
      <option
        value="value"
      >
        value
      </option>
      <option
        value=""
      />
    </select>
    <div
      class="invalid-feedback"
    >
      Invalid scan mode
    </div>
  </div>
</div>
`;

exports[`OIbScanMode check ScanMode with value="value" and help 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class=""
      for="name"
    >
      label
    </label>
    <select
      class="oi-form-input is-invalid form-control"
      id="name"
      name="name"
    >
      <option
        value="value"
      >
        value
      </option>
      <option
        value=""
      />
    </select>
    <div
      class="invalid-feedback"
    >
      Invalid scan mode
    </div>
    <small
      class="form-text text-muted"
    >
      <div>
        help text
      </div>
    </small>
  </div>
</div>
`;
